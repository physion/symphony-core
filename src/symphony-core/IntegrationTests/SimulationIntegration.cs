using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using System.Xml;
using ApprovalTests;
using ApprovalTests.Reporters;
using HDF5DotNet;
using Symphony.Core;
using Symphony.SimulationDAQController;
using NUnit.Framework;
using IntegrationTests.Properties;

namespace IntegrationTests
{
    [TestFixture]
    class SimulationIntegration
    {
        private const short SIMULATION_TIMESTEP = 500;

        [Test]
        public void SingleEpoch(
            [Values(1000, 5000, 10000, 20000)] double sampleRate,
            [Values(1, 2)] int nChannels
            )
        {
            RunSingleEpoch(sampleRate, nChannels, new FakeEpochPersistor());
        }

        private void RunSingleEpoch(double sampleRate, int nChannels, EpochPersistor epochPersistor)
        {
            Epoch e;
            IExternalDevice dev0;
            RenderedStimulus stim1;
            IExternalDevice dev1;
            RenderedStimulus stim2;
            IList<IMeasurement> stimData;
            var controller = SetupController(sampleRate, out e, out dev0, out stim1, out dev1, out stim2, out stimData, nChannels);

            controller.RunEpoch(e, epochPersistor);

            Assert.AreEqual((TimeSpan)stim1.Duration, e.Responses[dev0].Duration);
            if (nChannels > 1)
                Assert.AreEqual((TimeSpan)stim2.Duration, e.Responses[dev1].Duration);

            var inputData = e.Responses[dev0].Data;
            const double MAX_VOLTAGE_DIFF = 0.001;
            int failures = inputData.Select((t, i) => t.QuantityInBaseUnit - stimData[i].QuantityInBaseUnit)
                .Count(dif => Math.Abs(dif) > (decimal) MAX_VOLTAGE_DIFF);

            Assert.AreEqual(0, failures);
        }

        [Test]
        [UseReporter(typeof(NUnitReporter))]
        public void SingleEpochXMLPersistence()
        {
            var sb = new System.Text.StringBuilder();
            var sxw = XmlWriter.Create(sb);
            var gID = new Guid("a5839fe9-90ef-4e39-bf26-8f75048306a3");
            var exp = new EpochXMLPersistor(sxw, () => gID);

            var g1ID = new Guid("a5839fe9-90ef-4e39-bf26-8f75048306f3");
            var startTime1 = new DateTimeOffset(2011, 8, 22, 11, 12, 0, 0, TimeSpan.FromHours(-6));

            exp.BeginEpochGroup("label", "source", new string[0], new Dictionary<string, object>(), g1ID, startTime1);
            RunSingleEpoch(5, 2, exp);
            exp.EndEpochGroup();

            exp.Close();

            Approvals.VerifyXml(sb.ToString());

        }


        [Test]
        //[UseReporter(typeof(FileLauncherReporter))]
        public void SingleEpochHDF5Persistence()
        {
            if (File.Exists("SingleEpochHDF5Persistence.h5"))
                File.Delete("SingleEpochHDF5Persistence.h5");

            var gID = new Guid("a5839fe9-90ef-4e39-bf26-8f75048306a4");
            using (var exp = new EpochHDF5Persistor("SingleEpochHDF5Persistence.h5", null, () => gID))
            {
                var time = new DateTimeOffset(2011, 8, 22, 11, 12, 0, 0, TimeSpan.FromHours(-6));
                var guid = new Guid("053C64D4-ED7A-4128-AA43-ED115716A97D");

                exp.BeginEpochGroup("label1",
                    "source1",
                    new string[0],
                    new Dictionary<string, object>(),
                    guid,
                    time);

                RunSingleEpoch(5000, 2, exp);
                exp.EndEpochGroup();
                
                exp.Close();
            }

            Approvals.VerifyFile("SingleEpochHDF5Persistence.h5");

        }

        [Test]
        //[UseReporter(typeof(FileLauncherReporter))]
        public void AppendToExistingHDF5()
        {
            if (File.Exists("AppendToExistingHDF5.h5"))
                File.Delete("AppendToExistingHDF5.h5");

            var gID = new Guid("a5839fe9-90ef-4e39-bf26-8f75048306a4");
            using (var exp = new EpochHDF5Persistor("AppendToExistingHDF5.h5", null, () => gID))
            {
                var time = new DateTimeOffset(2011, 8, 22, 11, 12, 0, 0, TimeSpan.FromHours(-6));
                var guid = new Guid("053C64D4-ED7A-4128-AA43-ED115716A97D");

                exp.BeginEpochGroup("label1",
                    "source1",
                    new string[0],
                    new Dictionary<string, object>(),
                    guid,
                    time);

                RunSingleEpoch(5000, 2, exp);
                exp.EndEpochGroup();

                exp.Close();
            }

            gID = new Guid("a5839fe9-90ef-4e39-bf26-8f75048306a5");
            using (var exp = new EpochHDF5Persistor("AppendToExistingHDF5.h5", null, () => gID))
            {
                var time = new DateTimeOffset(2011, 8, 22, 11, 12, 0, 0, TimeSpan.FromHours(-6));
                var guid = new Guid("053C64D4-ED7A-4128-AA43-ED115716A97E");

                exp.BeginEpochGroup("label1",
                    "source1",
                    new string[0],
                    new Dictionary<string, object>(),
                    guid,
                    time);

                RunSingleEpoch(5000, 2, exp);
                exp.EndEpochGroup();

                exp.Close();
            }


            Approvals.VerifyFile("AppendToExistingHDF5.h5");
        }


        [Test]
        public void MultipleEpochs(
            [Values(1000, 5000, 10000, 20000)] double sampleRate
            )
        {
            Epoch e;
            IExternalDevice dev0;
            RenderedStimulus stim1;
            IExternalDevice dev1;
            RenderedStimulus stim2;
            IList<IMeasurement> stimData;
            var controller = SetupController(sampleRate, out e, out dev0, out stim1, out dev1, out stim2, out stimData, 2);

            controller.RunEpoch(e, new FakeEpochPersistor());

            var inputData = e.Responses[dev0].Data;
            const double MAX_VOLTAGE_DIFF = 0.001;
            int failures = inputData.Select((t, i) => t.QuantityInBaseUnit - stimData[i].QuantityInBaseUnit)
                .Count(dif => Math.Abs(dif) > (decimal) MAX_VOLTAGE_DIFF);

            Assert.AreEqual(0, failures);

            e = new Epoch("LowGainSimulation");
            dev0 = controller.GetDevice("Device0");
            dev1 = controller.GetDevice("Device1");

            var srate = new Measurement((decimal) sampleRate, "Hz");

            stim1 = new RenderedStimulus((string) "RenderedStimulus",
                                         (IDictionary<string, object>) new Dictionary<string, object>(),
                                         (IOutputData) new OutputData(stimData, srate, false));
            stim2 = new RenderedStimulus((string) "RenderedStimulus",
                                         (IDictionary<string, object>) new Dictionary<string, object>(),
                                         (IOutputData) new OutputData(stimData, srate, false));

            e.Stimuli[dev0] = stim1;
            e.Stimuli[dev1] = stim2;

            e.Responses[dev0] = new Response();
            e.Responses[dev1] = new Response();

            e.Background[dev1] = new Epoch.EpochBackground(new Measurement(0, "V"), srate);
            e.Background[dev0] = new Epoch.EpochBackground(new Measurement(0, "V"), srate);

            Assert.DoesNotThrow(() => controller.RunEpoch(e, new FakeEpochPersistor()));

            inputData = e.Responses[dev0].Data;
            failures = inputData.Select((t, i) => t.QuantityInBaseUnit - stimData[i].QuantityInBaseUnit)
                .Count(dif => Math.Abs(dif) > (decimal) MAX_VOLTAGE_DIFF);


            Assert.AreEqual(0, failures);

        }

        private static Controller SetupController(double sampleRate, out Epoch e, out IExternalDevice dev0, out RenderedStimulus stim1, out IExternalDevice dev1, out RenderedStimulus stim2, out IList<IMeasurement> stimData, int nChannels)
        {
            Converters.Clear();
            Converters.Register("V", "V",
                // just an identity conversion for now, to pass Validate()
                                (IMeasurement m) => m);

            var streamNameMap = new Dictionary<string, string>();
            streamNameMap["Out0"] = "In0";
            if (nChannels > 1)
                streamNameMap["Out1"] = "In1";

            Controller controller = new Parser().ParseConfiguration(Resources.LowGainConfig);

            var daq = (SimulationDAQController)controller.DAQController;
            daq.Clock = daq;
            foreach (var stream in daq.Streams)
            {
                stream.SampleRate = new Measurement((decimal) sampleRate, "Hz");
            }

            daq.SimulationRunner += (output, timestep) =>
                                        {
                                            var input = new ConcurrentDictionary<IDAQInputStream, IInputData>();

                                            Parallel.ForEach(output, (kv) =>
                                                                         {
                                                                             var outData = kv.Value;
                                                                             var outStream = kv.Key;
                                                                             var inStream = daq.InputStreams.Where((s) =>
                                                                                                                   s.Name == streamNameMap[outStream.Name]).First();

                                                                             var data = outData.DataWithUnits("V").Data;
                                                                             var inData = new InputData(data,
                                                                                                        outData.SampleRate,
                                                                                                        DateTimeOffset.Now)
                                                                                                        .DataWithNodeConfiguration("SimulationController",daq.Configuration);

                                                                             input[inStream] = inData;
                                                                         }
                                                );

                                            return input;
                                        };


            var protocolParams = new Dictionary<string, object>(1);
            protocolParams["key1"] = "value1";

            e = new Epoch("LowGainSimulation", protocolParams);
            dev0 = controller.GetDevice("Device0");
            dev1 = controller.GetDevice("Device1");

            if (nChannels == 1)
            {
                dev1.UnbindStream(dev1.Streams.Values.First().Name);
            }

            stimData = Enumerable.Range(0, (int)(10 * sampleRate))
                                   .Select(i => new Measurement(i, -3, "V") as IMeasurement)
                                   .ToList();
            var srate = new Measurement((decimal) sampleRate, "Hz");

            stim1 = new RenderedStimulus((string) "RenderedStimulus",
                                         (IDictionary<string, object>) new Dictionary<string, object>(),
                                         (IOutputData) new OutputData(stimData, srate, false));
            stim2 = new RenderedStimulus((string) "RenderedStimulus",
                                         (IDictionary<string, object>) new Dictionary<string, object>(),
                                         (IOutputData) new OutputData(stimData, srate, false));

            e.Stimuli[dev0] = stim1;

            if (nChannels > 1)
                e.Stimuli[dev1] = stim2;

            e.Responses[dev0] = new Response();
            if (nChannels > 1)
                e.Responses[dev1] = new Response();

            e.Background[dev1] = new Epoch.EpochBackground(new Measurement(0, "V"), srate);
            e.Background[dev0] = new Epoch.EpochBackground(new Measurement(0, "V"), srate);

            return controller;
        }
    }
}
