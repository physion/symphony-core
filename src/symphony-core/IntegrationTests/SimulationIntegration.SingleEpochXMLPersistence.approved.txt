<?xml version="1.0" encoding="utf-16"?><experiment><epochGroup label="label" identifier="a5839fe9-90ef-4e39-bf26-8f75048306f3" startTime="8/22/2011 5:12:00 PM +00:00" timeZoneUTCOffsetHours="-6" source="source"><keywords /><properties /><epoch protocolID="LowGainSimulation" UUID="a5839fe9-90ef-4e39-bf26-8f75048306a3" startTime="10/5/2011 4:21:31 PM -04:00"><background><Device1><backgroundMeasurement><measurement qty="0" unit="V" /></backgroundMeasurement><sampleRate><measurement qty="5" unit="Hz" /></sampleRate></Device1><Device0><backgroundMeasurement><measurement qty="0" unit="V" /></backgroundMeasurement><sampleRate><measurement qty="5" unit="Hz" /></sampleRate></Device0></background><protocolParameters><key1>value1</key1></protocolParameters><keywords /><stimuli><stimulus device="Device0" stimulusID="RenderedStimulus"><parameters /></stimulus><stimulus device="Device1" stimulusID="RenderedStimulus"><parameters /></stimulus></stimuli><responses><response device="Device0"><inputTime>10/5/2011 8:21:31 PM +00:00</inputTime><sampleRate><measurement qty="5" unit="Hz" /></sampleRate><nodeConfigurationSpans><nodeConfigurationSpan startTimeSeconds="0" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="0.6" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="1.2" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="1.8" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="2.4" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="3" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="3.6" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="4.2" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="4.8" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="5.4" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="6" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="6.6" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="7.2" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="7.8" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="8.4" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="9" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="9.6" timeSpanSeconds="0.4"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan></nodeConfigurationSpans><data><measurement qty="0" unit="V" /><measurement qty="1" unit="V" /><measurement qty="2" unit="V" /><measurement qty="3" unit="V" /><measurement qty="4" unit="V" /><measurement qty="5" unit="V" /><measurement qty="6" unit="V" /><measurement qty="7" unit="V" /><measurement qty="8" unit="V" /><measurement qty="9" unit="V" /><measurement qty="10" unit="V" /><measurement qty="11" unit="V" /><measurement qty="12" unit="V" /><measurement qty="13" unit="V" /><measurement qty="14" unit="V" /><measurement qty="15" unit="V" /><measurement qty="16" unit="V" /><measurement qty="17" unit="V" /><measurement qty="18" unit="V" /><measurement qty="19" unit="V" /><measurement qty="20" unit="V" /><measurement qty="21" unit="V" /><measurement qty="22" unit="V" /><measurement qty="23" unit="V" /><measurement qty="24" unit="V" /><measurement qty="25" unit="V" /><measurement qty="26" unit="V" /><measurement qty="27" unit="V" /><measurement qty="28" unit="V" /><measurement qty="29" unit="V" /><measurement qty="30" unit="V" /><measurement qty="31" unit="V" /><measurement qty="32" unit="V" /><measurement qty="33" unit="V" /><measurement qty="34" unit="V" /><measurement qty="35" unit="V" /><measurement qty="36" unit="V" /><measurement qty="37" unit="V" /><measurement qty="38" unit="V" /><measurement qty="39" unit="V" /><measurement qty="40" unit="V" /><measurement qty="41" unit="V" /><measurement qty="42" unit="V" /><measurement qty="43" unit="V" /><measurement qty="44" unit="V" /><measurement qty="45" unit="V" /><measurement qty="46" unit="V" /><measurement qty="47" unit="V" /><measurement qty="48" unit="V" /><measurement qty="49" unit="V" /></data></response><response device="Device1"><inputTime>10/5/2011 8:21:31 PM +00:00</inputTime><sampleRate><measurement qty="5" unit="Hz" /></sampleRate><nodeConfigurationSpans><nodeConfigurationSpan startTimeSeconds="0" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="0.6" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="1.2" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="1.8" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="2.4" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="3" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="3.6" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="4.2" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="4.8" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="5.4" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="6" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="6.6" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="7.2" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="7.8" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="8.4" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="9" timeSpanSeconds="0.6"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan><nodeConfigurationSpan startTimeSeconds="9.6" timeSpanSeconds="0.4"><SimulationController><Clock>Clock</Clock></SimulationController><STREAM><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></STREAM><EXTERNAL_DEVICE><MeasurementConversionTarget>V</MeasurementConversionTarget><Clock>Clock</Clock></EXTERNAL_DEVICE></nodeConfigurationSpan></nodeConfigurationSpans><data><measurement qty="0" unit="V" /><measurement qty="1" unit="V" /><measurement qty="2" unit="V" /><measurement qty="3" unit="V" /><measurement qty="4" unit="V" /><measurement qty="5" unit="V" /><measurement qty="6" unit="V" /><measurement qty="7" unit="V" /><measurement qty="8" unit="V" /><measurement qty="9" unit="V" /><measurement qty="10" unit="V" /><measurement qty="11" unit="V" /><measurement qty="12" unit="V" /><measurement qty="13" unit="V" /><measurement qty="14" unit="V" /><measurement qty="15" unit="V" /><measurement qty="16" unit="V" /><measurement qty="17" unit="V" /><measurement qty="18" unit="V" /><measurement qty="19" unit="V" /><measurement qty="20" unit="V" /><measurement qty="21" unit="V" /><measurement qty="22" unit="V" /><measurement qty="23" unit="V" /><measurement qty="24" unit="V" /><measurement qty="25" unit="V" /><measurement qty="26" unit="V" /><measurement qty="27" unit="V" /><measurement qty="28" unit="V" /><measurement qty="29" unit="V" /><measurement qty="30" unit="V" /><measurement qty="31" unit="V" /><measurement qty="32" unit="V" /><measurement qty="33" unit="V" /><measurement qty="34" unit="V" /><measurement qty="35" unit="V" /><measurement qty="36" unit="V" /><measurement qty="37" unit="V" /><measurement qty="38" unit="V" /><measurement qty="39" unit="V" /><measurement qty="40" unit="V" /><measurement qty="41" unit="V" /><measurement qty="42" unit="V" /><measurement qty="43" unit="V" /><measurement qty="44" unit="V" /><measurement qty="45" unit="V" /><measurement qty="46" unit="V" /><measurement qty="47" unit="V" /><measurement qty="48" unit="V" /><measurement qty="49" unit="V" /></data></response></responses></epoch></epochGroup></experiment>